### DISCUSSION TITLE
Implementing a constraint on when a vehicle has to return to a depot with a delivery

### DISCUSSION THREAD

**Discussion by choward47** [google/or-tools#2418] [Routing (and legacy CP) questions]
**What language and solver does this apply to?**
Routing

**Describe the problem you are trying to solve.**
I working on a vehicle routing problem and am trying to implement a constraint that forces a vehicle to return to the depot by the delivery time requirements for the nodes visited.

Below is what I am currently trying to use to implement this constraint. The code not included is framed from the example problems in the documentation. I'd be happy to elaborate more if needed/show other parts of the code.

```python
for vehicle_id in range(data['num_vehicles']):
    index = routing.Start(vehicle_id)
    for location_idx, delivery_window in enumerate(data['delivery window']):
        if location_idx >= len(df['Record_id']):
            continue
        if delivery_window is not None:
            if routing.VehicleVar(index) == routing.VehicleVar(location_idx):
                idx = routing.End(vehicle_id)
                solver.Add(time_dimension.CumulVar(idx) <= delivery_window)
```

The solver doesn't seem to be a fan of how I am trying to implement the constraint. Currently, it is not returning a solution despite the model having solutions before this constraint. Based on the data I am testing it on, it should be possible provided the constraint is implemented properly.

To summarize, I am asking for help in understanding how to go about writing a constraint that requires a vehicle returns to the depot before the delivery deadlines of the orders it has picked up. I was wondering if there is a functionality that already handles this that I have missed, or if some functionality could be created to achieve this.

Any help would be greatly appreciated! I've tried a few different things but I am reaching my wit's end.

**Comment by Mizux** [comment]
`routing.VehicleVar(index) == routing.VehicleVar(location_idx):`
will generate a `IntExpr` which is always evaluated to true by the python interpreter, i.e. you **can't** use it in a `if` statement **BUT** you can add it to the `solver.Add()` call in this case it will be evaluated by the routing solver each time a solution is produced to verify the constraint is not violated.

```python
idx = routing.End(vehicle_id)
same_vehicle = routing.VehicleVar(index) == routing.VehicleVar(location_idx)
routing.solver().Add(same_vehicle * time_dimension.CumulVar(idx) <= delivery_window)
```
If the constraint `same_vehicle` is evaluated to `false` aka 0 by the routing solver then the expression will become
`0 * time_dimension.CumulVar(idx) <= delivery_window` i.e. `0 <= delivery_window` which is OK

If the constraint `same_vehicle` is evaluated to `true` aka 1 by the routing solver then the expression will become
`1 * time_dimension.CumulVar(idx) <= delivery_window` i.e. `time_dimension.CumulVar(idx) <= delivery_window` which is also OK
